### PR TITLE
Add Scheduler

### PR DESCRIPTION
**Pull request by rentainhe**
## TODO List
添加可能用到的Scheduler，并且与注册机制兼容，统一命名标准`xxxLR`，每个单词首字母大写，比较规范整齐
- [x] WarmupCosineLR
- [x] WarmupMultiStepLR
- [ ] WarmupPolynomialLR（需要等PolynomialLR加入到oneflow中）
- [x] WarmupFixedStepLR
- [x] WarmupCosineAnnealingLR
- [x] 将LRScheduler中的cfg参数拆分出来，这样也可以支持`LazyCall`或者其他project使用
- [x] 统一config中的样式，以便可以使用一个`build_lr_scheduler`函数统一定义scheduler
- [x] 添加`Warning`信息，用来判断warmup_iter和max_iter的关系
- [x] 添加到`test_trainer`中，进行测试
- [x] 添加`lr_scheduler`测试

## Problem needed to fix
- [ ] WarmupLR Scheduler更新问题: https://github.com/Oneflow-Inc/libai/issues/37

与这个PR https://github.com/Oneflow-Inc/libai/pull/26 内容一样，原 PR https://github.com/Oneflow-Inc/libai/pull/26 仅仅用于测试，不合并

## Example
- `build_lr_scheduler` func
```python
def build_lr_scheduler(cfg, optimizer):
    """ Build learning rate scheduler, defined by ``cfg.optim.lr_scheduler``.
    """
    if "_target_" in cfg:
        cfg.optimizer = optimizer
        scheduler = instantiate(cfg)
    else:
        scheduler_name = cfg.name
        del cfg.name
        scheduler = SCHEDULER_REGISTRY.get(scheduler_name)(optimizer, **cfg)
    return scheduler
```
- 支持注册机制，兼容传入config和optimizer，得到对应的lr_scheduler

这里需要统一一个命名规范，传入的cfg中必须包含`scheduler_name`和`scheduler_cfg`，表示当前使用的lr_scheduler是什么，然后接对应的参数
```python
# config.py 里定义了一个这样的config
lr_scheduler = dict(
    scheduler_name = "WarmupCosineLR",
    scheduler_cfg = dict(
        max_iters = 2000,
        alpha = 0.001,
        warmup_factor = 0.001,
        warmup_iters = 1000,
        warmup_method = "linear"
    )
)
```
```python
# 可以在别的文件中按如下方法调用
import oneflow as flow
from flowvision.models.resnet import resnet18
from libai.scheduler import build_lr_scheduler
from libai.config import LazyConfig

cfg = LazyConfig.load("./config.py")
model = resnet18()
optim = flow.optim.SGD(model.parameters(), 0.001)
# 加载config，传入定义好的optimizer
scheduler = build_lr_scheduler(cfg.lr_scheduler, optim)
```
- 支持LazyCall调用，直接定义一个`LazyCall`，和optim传入`build_lr_scheduler`
```python
import oneflow as flow
from libai.config import LazyCall, instantiate
from libai.scheduler import build_lr_scheduler, WarmupCosineLR
from flowvision.models.resnet import resnet18

lr_scheduler = LazyCall(WarmupCosineLR)(
    max_iters = 2000,
    alpha = 0.001,
    warmup_factor = 0.001,
    warmup_iters = 1000,
    warmup_method = "linear"
)

model = resnet18()
optim = flow.optim.SGD(model.parameters(), 0.001)
# 调用instantiate实例化这个lr scheduler
lr_scheduler = build_lr_scheduler(lr_scheduler, optim)
```
- 支持其他project使用libai的scheduler，例如可以在flowvision下使用
```python
import oneflow as flow
from libai.scheduler import WarmupCosineLR
from flowvision.models.resnet import resnet18



model = resnet18()
optim = flow.optim.SGD(model.parameters(), 0.001)
lr_scheduler = WarmupCosineLR(optimizer=optim, 
                              max_iters=2000, 
                              alpha=0.001, 
                              warmup_factor=0.001, 
                              warmup_iters=1000,
                              warmup_method="linear")
```

### CODE REVIEW

File: libai/scheduler/build.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+The registered object will be called with `obj(cfg)` 
+and expected to return a `flow.optim.lr_scheduler._LRScheduler` object.
+"""
+
+def build_lr_scheduler(cfg, optimizer):
+    """ Build learning rate scheduler, defined by ``cfg.optim.lr_scheduler``.
+    """
+    if "_target_" in cfg:
+        cfg.optimizer = optimizer
```
Comment (rentainhe) [inline libai/scheduler/build.py:31]: 在这一步添加optimizer参数

File: libai/scheduler/build.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        scheduler = instantiate(cfg)
+    else:
+        scheduler_name = cfg.name
+        del cfg.name
```
Comment (rentainhe) [inline libai/scheduler/build.py:35]: 需要手动delete name这个args，可以看看是否需要再简化，因为我的想法是尽量简化config，而不是build_lr_scheduler函数，这样可能使用起来更加简单一些，需要考虑的config格式会比较少

File: libai/scheduler/lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+import oneflow as flow
+from oneflow.nn.optimizer.lr_scheduler import LrScheduler
+
+from libai.config.config import configurable
+
+from .build import SCHEDULER_REGISTRY
+
+
+class PolynomialLR(LrScheduler):
```
Comment (L1aoXingyu) [inline libai/scheduler/lr_scheduler.py:26]: 这个删掉吧，我们直接合并到 oneflow 主库里面

File: libai/scheduler/build.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+Registry for lr scheduler, i.e. WarmupCosineLR
+
+The registered object will be called with `obj(cfg)` 
+and expected to return a `flow.optim.lr_scheduler._LRScheduler` object.
+"""
+
+def build_lr_scheduler(cfg, optimizer):
+    """ Build learning rate scheduler, defined by ``cfg.optim.lr_scheduler``.
+    """
+    if "_target_" in cfg:
```
Comment (L1aoXingyu) [inline libai/scheduler/build.py:30]: 整个 config.py 在使用的时候是 cfg，里面每一个参数都是在 cfg 下的 namespace，所以这里应该是 `cfg.lr_scheduler`

**Review by L1aoXingyu — COMMENTED** [review]
可以在 configs/common 里面增加 optimizer 和 lr scheduler 的默认配置

File: libai/scheduler/build.py (hunk 2 above)
Comment (L1aoXingyu) [reply libai/scheduler/build.py:35]: 这里不删掉有什么后果吗，不太明白这个地方的意义，可以再解释一下

File: libai/scheduler/build.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from libai.config import instantiate
+from libai.utils.registry import Registry
+
+SCHEDULER_REGISTRY = Registry("Scheduler")
```
Comment (CPFLAME) [inline libai/scheduler/build.py:19]: 有个问题， 既然我们的lr_schduler是在lazyconfig中指定到函数名去延迟初始化的，
那我们为什么还需要register？

File: libai/scheduler/lr_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            learning_rate_decay_conf.mutable_polynomial_conf().set_power(self.power)
+            learning_rate_decay_conf.mutable_polynomial_conf().set_cycle(self.cycle)
+
+
+
+@SCHEDULER_REGISTRY.register()
+def WarmupCosineLR(optimizer: flow.optim.Optimizer,
+                   max_iters: int,
+                   warmup_factor: float,
+                   warmup_iters: int,
```
Comment (CPFLAME) [inline libai/scheduler/lr_scheduler.py:106]: 这里warmup_iters可以判断一下 是否超过总的iter， 如果超过了 可以打印一个信息

File: libai/scheduler/build.py (hunk 5 above)
Comment (L1aoXingyu) [reply libai/scheduler/build.py:19]: 这是为了兼容两种用户习惯，让不熟悉 lazyCall 的用户可以用他熟悉的注册方式去调用对应的对象

File: libai/scheduler/build.py (hunk 2 above)
Comment (rentainhe) [reply libai/scheduler/build.py:35]: 这里确实有点问题，会改一下，统一一下cfg规范

File: libai/scheduler/lr_scheduler.py (hunk 3 above)
Comment (rentainhe) [reply libai/scheduler/lr_scheduler.py:26]: 好，那暂时先不加这个，等主库更新后再加进来

File: libai/scheduler/build.py (hunk 5 above)
Comment (CPFLAME) [reply libai/scheduler/build.py:19]: 那其他的模块，比如 model，optimizer 等是不是也应该兼容两种用户习惯
不然用户在lr_scheduler用register， 在model用lazycall， 会很别扭

File: libai/scheduler/build.py (hunk 4 above)
Comment (rentainhe) [reply libai/scheduler/build.py:30]: 好，我想的是可以从外部传进来
```python
build_lr_scheduler(cfg.lr_scheduler, optimizer)
```

File: libai/scheduler/build.py (hunk 5 above)
Comment (L1aoXingyu) [reply libai/scheduler/build.py:19]: 对，现在 model 也支持了两种方式

Comment (rentainhe) [reply libai/scheduler/build.py:19]: 可以兼容一下感觉，因为也不太影响LazyCall的使用

File: libai/scheduler/lr_scheduler.py (hunk 6 above)
Comment (rentainhe) [reply libai/scheduler/lr_scheduler.py:106]: OK，我加到TODO

File: libai/scheduler/build.py (hunk 5 above)
Comment (CPFLAME) [reply libai/scheduler/build.py:19]: 可以的

File: libai/scheduler/build.py (hunk 4 above)
Comment (L1aoXingyu) [reply libai/scheduler/build.py:30]: 一般的使用场景是这样
```python
class MyTrainer(DefaultTrainer):
	
	@classmethod
	def buidl_lr_scheduler(cfg, optimizer):
		# do some customized computing
		# 这里可能会用到 cfg.train cfg.model 里面的某些参数，
		# 所以需要传入完整的 cfg
		return build_lr_scheduler(cfg, optimizer)
```
因为 build_lr_scheduler 的函数定义里面只有 cfg，所以用户会理所当然觉得直接传入 cfg 就可以，这时候就和 defaultTrainer 里面的行为不一样了，所以我建议要不在 build_lr_scheduler 的函数定义入参不要定义成 cfg，定义成 cfg_scheduler 之类的，要不在 build_lr_scheduler 内部做一层解析

File: libai/scheduler/lr_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                   alpha: float = 0.0,
+                   warmup_method: str = "linear",
+                   **kwargs):
+    cosine_decay_lr = flow.optim.lr_scheduler.CosineDecayLR(
+        optimizer, decay_steps=max_iters, alpha=alpha
+    )
+    if warmup_iters == 0:
+        logger.warning("warmup iters equals to zero, directly return the passed scheduler")
+        return cosine_decay_lr
+    elif warmup_iters > max_iters:
```
Comment (rentainhe) [inline libai/scheduler/lr_scheduler.py:38]: 在这里做了一层判断，打印一个warning信息

File: libai/scheduler/lr_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+@SCHEDULER_REGISTRY.register()
+def WarmupCosineAnnealingLR(optimizer: flow.optim.Optimizer,
+                            t_max: int,
+                            warmup_factor: float,
+                            warmup_iters: int,
+                            eta_min: float = 0.0,
+                            warmup_method: str = "linear",
+                            **kwargs):
+    cosine_annealing_lr = flow.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=t_max, eta_min=eta_min)
+    if warmup_iters == 0:
```
Comment (rentainhe) [inline libai/scheduler/lr_scheduler.py:55]: 判断warmup_iters如果等于0的话，直接返回schduler

**Review by L1aoXingyu — APPROVED** [review]
